dask/array/core.py

```diff
@@ -232,46 +233,88 @@ def slices_from_chunks(chunks):
     return list(product(*slices))
 
 
-def getem(
-    arr,
+def graph_from_arraylike(
+    arr,  # Any array-like which supports slicing
```

Comment:
could use a `Protocol` 😉 
In general, type annotations on this function would be nice for readability

Reply:
I think we can't use `Protocol` yet due to needing `typing_extensions==3.10`, which has a few known incompatibilities. At least it did the last time I checked a few months ago

Reply:
Ah, wait, I was thinking of `ParamSpec`

Reply:
Took a pass at this, but I'm afraid it's still of limited utility, since going much further would have to answer things like "What is the Array protocol anyways?", and "what kinds of things can `normalize_chunks` take?"

Reply:
Yeah, what you added already feels very helpful, but it's unfortunate that it's so hard to go beyond that.